pkg/daemon/ceph/osd/volume.go

```diff
@@ -0,0 +1,214 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package osd
+
+import (
+	"encoding/json"
+	"fmt"
+	"path"
+	"strconv"
+	"syscall"
+
+	"github.com/rook/rook/pkg/clusterd"
+	oposd "github.com/rook/rook/pkg/operator/ceph/cluster/osd"
+	"github.com/rook/rook/pkg/operator/ceph/cluster/osd/config"
+	"github.com/rook/rook/pkg/util/exec"
+)
+
+var cephConfigDir = "/var/lib/ceph"
+
+const (
+	osdsPerDeviceFlag = "--osds-per-device"
+	encryptedFlag     = "--dmcrypt"
+	cephVolumeCmd     = "ceph-volume"
+)
+
+func (a *OsdAgent) configureDevices(context *clusterd.Context, devices *DeviceOsdMapping) ([]oposd.OSDInfo, bool, error) {
+	var osds []oposd.OSDInfo
+	if a.metadataDevice != "" {
+		logger.Infof("skipping ceph-volume until the fast devices can be specified for the metadata")
+		return osds, false, nil
+	}
+
+	useCephVolume, err := getCephVolumeSupported(context)
+	if err != nil {
+		return nil, false, fmt.Errorf("failed to detect if ceph-volume is available. %+v", err)
+	}
+	if !useCephVolume {
+		return osds, false, nil
+	}
+
+	if devices == nil || len(devices.Entries) == 0 {
+		logger.Infof("no new devices to configure. returning devices already configured with ceph-volume.")
+		osds, err = getCephVolumeOSDs(context, a.cluster.Name)
+		if err != nil {
+			logger.Infof("failed to get devices already provisioned by ceph-volume. %+v", err)
+		}
+		return osds, true, nil
+	}
+
+	err = createOSDBootstrapKeyring(context, a.cluster.Name, cephConfigDir)
+	if err != nil {
+		return nil, true, fmt.Errorf("failed to generate osd keyring. %+v", err)
+	}
+
+	if err = a.initializeDevices(context, devices); err != nil {
+		return nil, true, fmt.Errorf("failed to initialize devices. %+v", err)
+	}
+
+	osds, err = getCephVolumeOSDs(context, a.cluster.Name)
+	return osds, true, err
+}
+
+func (a *OsdAgent) initializeDevices(context *clusterd.Context, devices *DeviceOsdMapping) error {
+	storeFlag := "--bluestore"
+	if a.storeConfig.StoreType == config.Filestore {
+		storeFlag = "--filestore"
+	}
+
+	baseArgs := []string{"lvm", "batch", "--prepare", storeFlag, "--yes"}
+	if a.storeConfig.EncryptedDevice {
+		baseArgs = append(baseArgs, encryptedFlag)
+	}
+
+	batchArgs := append(baseArgs, []string{
+		osdsPerDeviceFlag,
+		strconv.Itoa(a.storeConfig.OSDsPerDevice),
+	}...)
+
+	// ceph-volume is soon implementing a parameter to specify the "fast devices", which correspond to the "metadataDevice" from the
+	// crd spec. After that is implemented, we can implement this. In the meantime, we fall back to use rook's partitioning.
+	metadataDeviceSpecified := false
+
+	configured := 0
+	for name, device := range devices.Entries {
+		if device.LegacyPartitionsFound {
+			logger.Infof("skipping device %s configured with legacy rook osd", name)
+			continue
+		}
+
+		if device.Data == -1 {
```

Comment:
This comes from `ceph-volume lvm list` I suppose, right? Not 100% sure but I think when an OSD has been configured with ceph-volume filestore we will see `data` and for bluestore we will see `block` so we have to be careful not trying to prepare an existing osd.

Reply:
`device.Data == -1` is from Rook, indicating that the device hasn't been previously configured. Before we get to this point, Rook has called `ceph-volume` and looked at other legacy rook osds. If they have been configured the will have a non-negative osd id here. If it hasn't been configured it is -1.

Reply:
And I did notice that if filestore, the type will be `data` or `journal`, and bluestore will be `block`. Down on line 179 `journal` is used as the keyword to determine filestore.

Reply:
Ok thanks :)